File: Random/OperacionesArboles.go
Funcion buscar(prim: puntero a nodo, dato: entero):entero 
	si prim.dato = dato entonces 
		buscar:= dato 
	fsi 

	si prim = nil entonces 
		buscar:= nil 
	fsi 
	 
	si *prim.dato < dato entonces 
		buscar:= buscar(*prim.der,dato)
	sino 
		si *prim.dato > dato 
			buscar:= buscar(*prim.izq,dato)
		fsi 
	fsi 

FFuncion 

Procedimiento insertar(prim, q: puntero a nodo)
	si prim = nil entonces 
		*q.der:= nil 
		*q.izq:= nil 
		prim:= q 
	sino 
		insertarRecursivo(prim,q)
	fsi 
FProcedimiento 

Procedimiento insertarRecursivo(prim, q: puntero a nodo)
	si *q.dato <= *prim.dato entonces 
		si *prim.izq = nil entonces 
			*prim.izq:= q 
			*q.izq:= nil 
			*q.der:= nil 
		sino 
			insertarRecursivo(*prim.izq,q)
		fsi 
	sino 
		si *q.dato > *prim.dato entonces 
			si *prim.der = nil entonces
				*prim.der:= q 
				*q.izq:= nil 
				*q.der:= nil 
			sino 
				insertarRecursivo(*prim.der,q)
			fsi 
		fsi 
	fsi 
FProcedimiento

Funcion calcNivel(prim: puntero a nodo): entero 
	si prim = nil entonces 
		calcNivel:= 0 
	sino 
		si calcNivel(*prim.izq) > calcNivel(*prim.der) entonces 
			calcNivel:= 1 + calcNivel(*prim.izq)
		sino 
			calcNivel:= 1 + calcNivel(*prim.der)
		Fsi 
	Fsi 
FFuncion

Procedimiento enOrden(prim: puntero a nodo)
	si prim <> nil entonces 
		enOrden(*prim.izq)
		tratarRaiz()
		enOrden(*prim.der)
	fsi 
FProcedimiento

Procedimiento preOrden(prim: puntero a nodo)
	si prim <> nil entonces 
		tratarRaiz()
		preOrden(*prim.izq)
		preOrden(*prim.der)
	fsi 
FProcedimiento

Procedimiento postOrden(prim: puntero a nodo)
	si prim <> nil entonces 
		postOrden(*prim.izq)
		postOrden(*prim.der)
		tratarRaiz()
	fsi 
FProcedimiento

File: Random/OperacionesArboles.py
class Nodo():
    def __init__(self, numero):
        self.numero = numero 
        self.izq = None
        self.der = None 

class ABB():
    def __init__(self):
        self.prim = None 

    def cargaArbol(self, numero):
        if self.prim is None:
            self.prim = Nodo(numero)
        else:
            self.insertarRecursivo(self.prim, numero)

    def insertarRecursivo(self, p, numero):
        if numero < p.numero:
            if p.izq is None:
                p.izq = Nodo(numero)
            else:
                self.insertarRecursivo(p.izq, numero)
        elif numero > p.numero:
            if p.der is None:
                p.der = Nodo(numero)
            else:
                self.insertarRecursivo(p.der, numero)

def enOrden(nodo: ABB):
    if nodo is not None:
        enOrden(nodo.izq)
        print(f"{nodo.numero}",end=" -> ")
        enOrden(nodo.der)

def postOrden(nodo: ABB):
    if nodo is not None:
        postOrden(nodo.izq)
        postOrden(nodo.der)
        print(f"{nodo.numero}",end=" -> ")

def preOrden(nodo: ABB):
    if nodo is not None:
        print(f"{nodo.numero}",end=" -> ")
        preOrden(nodo.izq)
        preOrden(nodo.der)

def busqueda(prim: ABB, buscar:int) -> int:
    if prim is None:
        return -1

    if prim.numero == buscar:
        return prim.numero 

    if buscar < prim.numero:
        return busqueda(prim.izq,buscar)
    else:
        return busqueda(prim.der,buscar)

def altura(nodo: ABB) -> int:
    if nodo is None:
        return 0 
    else:
        if altura(nodo.izq) > altura(nodo.der):
            return 1 + altura(nodo.izq)
        else:
            return 1 + altura(nodo.der)

def camino(nodo: ABB,valor: int):    
    if nodo is not None:
        print(f"{nodo.numero}",end=" -> ")

        if valor > nodo.numero:
            camino(nodo.der,valor)
        else:
            if valor < nodo.numero:
                camino(nodo.izq,valor)

def nodos(nodo: ABB) -> int:
    if nodo is None:
        return 0 
    else:
        if nodo is not None:
            return 1 + nodos(nodo.izq) + nodos(nodo.der)
        


arbol = ABB()

arbol.cargaArbol(50)
arbol.cargaArbol(30)
arbol.cargaArbol(10)
arbol.cargaArbol(60)
arbol.cargaArbol(40)
arbol.cargaArbol(70)
arbol.cargaArbol(80)

print(f"La altura del arbol es: {altura(arbol.prim)}")
print("Recorrido en orden:")
enOrden(arbol.prim)
print("\nRecorrido post orden:")
postOrden(arbol.prim)
print("\nRecorrido pre orden:")
preOrden(arbol.prim)
print(f"\nValor: {busqueda(arbol.prim,6)}")

print(f"La cantidad de nodos del arbol es: {nodos(arbol.prim)}")

